crates/solvers/src/fault/position.rs

```diff
@@ -0,0 +1,328 @@
+//! The position module holds the [Position] trait and its implementations.
+
+/// The [Position] trait defines the interface of a generalized index within a binary tree.
+/// A "Generalized Index" is calculated as `2^{depth} + index_at_depth`.
+pub trait Position {
+    /// Returns the depth of the [Position] within the tree.
+    fn depth(&self) -> u64;
+    /// Returns the index at depth of the [Position] within the tree.
+    fn index_at_depth(&self) -> u64;
+    /// Returns the left child [Position] relative to the current [Position].
+    fn left(&self) -> Self;
+    /// Returns the right child [Position] relative to the current [Position].
+    fn right(&self) -> Self;
+    /// Returns the parent [Position] relative to the current [Position].
+    fn parent(&self) -> Self;
+    /// Returns the rightmost [Position] that commits to the same trace index as the current [Position].
+    fn right_index(&self, max_depth: u64) -> Self;
+    /// Returns the trace index that the current [Position] commits to.
+    fn trace_index(&self, max_depth: u64) -> u64;
+    /// Returns the relative [Position] for an attack or defense move against the current [Position].
+    fn make_move(&self, is_attack: bool) -> Self;
+}
+
+/// Computes a generalized index from a depth and index at depth.
+///
+/// ### Takes
+/// - `depth`: The depth of the generalized index.
+/// - `index_at_depth`: The index at depth of the generalized index.
+///
+/// ### Returns
+/// - `u128`: The generalized index: `2^{depth} + index_at_depth`.
+pub fn compute_gindex(depth: u8, index_at_depth: u64) -> u128 {
+    2u128.pow(depth as u32) + index_at_depth as u128
+}
+
+/// Implementation of the [Position] trait for the [std::u128] primitive type.
+impl Position for u128 {
+    fn depth(&self) -> u64 {
+        127 - self.leading_zeros() as u64
+    }
+
+    fn index_at_depth(&self) -> u64 {
+        (self - (1 << self.depth())) as u64
+    }
+
+    fn left(&self) -> Self {
+        self << 1
+    }
+
+    fn right(&self) -> Self {
+        (self << 1) | 1
+    }
+
+    fn parent(&self) -> Self {
+        self >> 1
+    }
+
+    fn right_index(&self, max_depth: u64) -> Self {
+        let remaining = max_depth - self.depth();
+        (self << remaining) | ((1 << remaining) - 1)
+    }
+
+    fn trace_index(&self, max_depth: u64) -> u64 {
+        self.right_index(max_depth).index_at_depth()
+    }
+
+    fn make_move(&self, is_attack: bool) -> Self {
+        ((!is_attack as u128) | self) << 1
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{compute_gindex, Position};
+
+    const MAX_DEPTH: u64 = 4;
+
+    #[test]
+    fn position_correctness_static() {
```

Comment:
All these can probs use a golang-style table test or even macro(?) to avoid duplicate code